### PR TITLE
types: fix unexpected NOT_NULL flags

### LINKED ISSUE #18488: NOT_NULL Flag incorrectly set
### 1. Minimal reproduce step (Required)

````
CREATE TABLE `test` (
            `iD` bigint(20) NOT NULL,
            `INT_TEST` int(11) DEFAULT NULL
          ) ENGINE=InnoDB DEFAULT CHARSET=utf8 ROW_FORMAT=DYNAMIC;
INSERT INTO test VALUES (2, 10), (3, NULL);

 SELECT DISTINCT count(*), id + int_test as res FROM test  GROUP BY res ORDER BY res;
````

### 2. What did you expect to see? (Required)

````
mysql>  SELECT DISTINCT count(*), id + int_test as res FROM test  GROUP BY res ORDER BY res;
+----------+-----+
| count(*) | res |
+----------+-----+
|        1 | NULL|
|        1 |  12 |
+----------+-----+
2 rows in set (0.01 sec)
````
### 3. What did you see instead (Required)

````
mysql>  SELECT DISTINCT count(*), id + int_test as res FROM test  GROUP BY res ORDER BY res;
+----------+-----+
| count(*) | res |
+----------+-----+
|        1 | NULL |
|        1 |  12 |
+----------+-----+
2 rows in set (0.01 sec)
````

### 4. Affected version (Required)

````
commit a9177fe846bf2638261f64514ab4ebd0155cc60c (HEAD -> master, origin/master, origin/HEAD)
Author: Boqin Qin <bobbqqin@bupt.edu.cn>
Date:   Mon Jul 6 19:28:12 2020 +0800
````
### 5. Root Cause Analysis

An extra space is printed.

### LINKED ISSUE #19025: JSON type has an unexpected NOT_NULL flag.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```mysql
create table t ( c_bit bit(10), c_int_d int, c_uint_d int unsigned, c_bigint_d bigint, c_ubigint_d bigint unsigned, c_float_d float, c_ufloat_d float unsigned, c_double_d double, c_udouble_d double unsigned, c_decimal decimal(6, 3), c_udecimal decimal(10, 3) unsigned, c_decimal_d decimal, c_udecimal_d decimal unsigned, c_datetime datetime(2), c_datetime_d datetime, c_time time(3), c_time_d time, c_date date, c_timestamp timestamp(4) DEFAULT CURRENT_TIMESTAMP(4), c_timestamp_d timestamp DEFAULT CURRENT_TIMESTAMP, c_char char(20), c_bchar char(20) binary, c_varchar varchar(20), c_bvarchar varchar(20) binary, c_text_d text, c_btext_d text binary, c_binary binary(20), c_varbinary varbinary(20), c_blob_d blob, c_set set('a', 'b', 'c'), c_enum enum('a', 'b', 'c'), c_json JSON, c_year year );
```

### 2. What did you expect to see? (Required)

```mysql
mysql> select c_json from t;
Field   1:  `c_json`
Catalog:    `def`
Database:   `test`
Table:      `t`
Org_table:  `t`
Type:       JSON
Collation:  binary (63)
Length:     4294967295
Max_length: 0
Decimals:   0
Flags:      BLOB BINARY 


0 rows in set (0.02 sec)
```

### 3. What did you see instead (Required)

```mysql
mysql> select c_json from t;
Field   1:  `c_json`
Catalog:    `def`
Database:   `test`
Table:      `t`
Org_table:  `t`
Type:       JSON
Collation:  binary (63)
Length:     4294967295
Max_length: 0
Decimals:   0
Flags:      NOT_NULL BINARY 


0 rows in set (0.01 sec)
```

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve?

Issue Number: 
related to #18921, and close #19025, close #18488 

Problem Summary:

It is introduced from https://github.com/pingcap/tidb/pull/9689

### What is changed and how it works?

Proposal: [xxx](url) 
```
mysql> select convert(null , JSON);
Field   1:  `convert(null , JSON)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       JSON
Collation:  utf8mb4_bin (46)
Length:     4194304
Max_length: 0
Decimals:   0
Flags:      NOT_NULL BINARY 


+----------------------+
| convert(null , JSON) |
+----------------------+
| NULL                 |
+----------------------+
1 row in set (0.00 sec)
```

What's Changed:

- Rewrite https://github.com/pingcap/tidb/pull/9689 for fixing https://github.com/pingcap/tidb/issues/9463

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Fix the unexpected NOT_NULL flag

### CODE REVIEW

**Comment by nullnotnil** [comment]
Does this also fix https://github.com/pingcap/tidb/issues/18488 by chance?

**Comment by SunRunAway** [comment]
> Does this also fix #18488 by chance?

Yes, the previous implementation adds `NOT NULL` flag when a `BINARY` flag appears.

File: expression/constant.go [hunk 1]
```diff
@@ -66,6 +69,20 @@ type Constant struct {
 	collationInfo
 }
 
+// Clone implements Expression interface.
+func (c *Constant) Clone() Expression {
+	con := &Constant{
+		Value:         c.Value,
+		once:          sync.Once{},
```
Comment (lzmhhh123) [inline expression/constant.go:76]: Once not clone?

File: expression/constant.go [hunk 2]
```diff
@@ -110,6 +121,12 @@ func (c *Constant) GetType() *types.FieldType {
 		types.DefaultParamTypeForValue(dt.GetValue(), tp)
 		return tp
 	}
+	if !c.Value.IsNull() {
```
Comment (lzmhhh123) [inline expression/constant.go:124]: We can check `!c.Value.IsNull() && c.RetType.Flag&mysql.NotNullFlag == 0`.

Comment (lzmhhh123) [reply expression/constant.go:124]: Then we can remove the `sync.Once`.

**Comment by SunRunAway** [comment]
```
[2020-12-03T05:57:05.517Z] FAIL: db_partition_test.go:1466: testIntegrationSuite7.TestAlterTableExchangePartition

[2020-12-03T05:57:05.517Z] 

[2020-12-03T05:57:05.517Z] db_partition_test.go:1567:

[2020-12-03T05:57:05.517Z]     tk.MustExec("alter table e6 exchange partition p0 with table e7")

[2020-12-03T05:57:05.517Z] /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:217:

[2020-12-03T05:57:05.517Z]     tk.c.Assert(err, check.IsNil, check.Commentf("sql:%s, %v, error stack %v", sql, args, errors.ErrorStack(err)))

[2020-12-03T05:57:05.517Z] ... value *errors.withStack = [ddl:1736]Tables have different definitions ("[ddl:1736]Tables have different definitions")

[2020-12-03T05:57:05.517Z] ... sql:alter table e6 exchange partition p0 with table e7, [], error stack [ddl:1736]Tables have different definitions

[2020-12-03T05:57:05.517Z] github.com/pingcap/errors.AddStack

[2020-12-03T05:57:05.517Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20201126102027-b0a155152ca3/errors.go:174

[2020-12-03T05:57:05.518Z] github.com/pingcap/errors.Trace

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20201126102027-b0a155152ca3/juju_adaptor.go:15

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/ddl.checkTableDefCompatible

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/ddl/ddl_api.go:3080

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/ddl.(*ddl).ExchangeTablePartition

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/ddl/ddl_api.go:3178

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/ddl.(*ddl).AlterTable

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/ddl/ddl_api.go:2390

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/executor.(*DDLExec).executeAlterTable

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/executor/ddl.go:372

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/executor.(*DDLExec).Next

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/executor/ddl.go:86

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/executor.Next

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/executor/executor.go:277

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/executor/adapter.go:524

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelay

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/executor/adapter.go:405

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/executor.(*ExecStmt).Exec

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/executor/adapter.go:355

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/session.runStmt

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/session/session.go:1317

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/session.(*session).ExecuteStmt

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/session/session.go:1261

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/util/testkit.(*TestKit).Exec

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:170

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/util/testkit.(*TestKit).MustExec

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:216

[2020-12-03T05:57:05.518Z] github.com/pingcap/tidb/ddl_test.(*testIntegrationSuite7).TestAlterTableExchangePartition

[2020-12-03T05:57:05.518Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/ddl/db_partition_test.go:1567

[2020-12-03T05:57:05.519Z] reflect.Value.call

[2020-12-03T05:57:05.519Z] 	/usr/local/go/src/reflect/value.go:460

[2020-12-03T05:57:05.519Z] reflect.Value.Call

[2020-12-03T05:57:05.519Z] 	/usr/local/go/src/reflect/value.go:321

[2020-12-03T05:57:05.519Z] github.com/pingcap/check.(*suiteRunner).forkTest.func1

[2020-12-03T05:57:05.519Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:850

[2020-12-03T05:57:05.519Z] github.com/pingcap/check.(*suiteRunner).forkCall.func1

[2020-12-03T05:57:05.519Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739

[2020-12-03T05:57:05.519Z] runtime.goexit

[2020-12-03T05:57:05.519Z] 	/usr/local/go/src/runtime/asm_amd64.s:1357

```

File: server/tidb_test.go [hunk 3]
```diff
@@ -929,17 +929,60 @@ func (ts *tidbTestSuite) TestSumAvg(c *C) {
...
-	c.Assert(dumpFlag(cols[0].Type, cols[0].Flag), Equals, expectFlag)
+	{
+		// issue #9689
+		rs, err := Execute(ctx, qctx, "select 1")
+		c.Assert(err, IsNil)
+		cols := rs.Columns()
+		c.Assert(len(cols), Equals, 1)
+		expectFlag := uint16(tmysql.NotNullFlag | tmysql.BinaryFlag)
+		c.Assert(dumpFlag(cols[0].Type, cols[0].Flag), Equals, expectFlag)
+	}
```
Comment (blacktear23) [inline server/tidb_test.go:944]: We should add another case: `select null;` and the expectFlag should be `uint16(tmysql.BinaryFlag)`

**Comment by xuyifangreeneyes** [comment]
#21307 can also be fixed by the pr.

**Comment by xuyifangreeneyes** [comment]
#21307 can also be fixed by the pr.

**Comment by SunRunAway** [comment]
I'm trying to let `Constant.RetType` have a Not Null Flag to solve the entire issue of NOT_NULL flags.
But I was not able to solve the concurrency issue of setting `Constant.RetType`, some tests are broken. I have no time to continue to fix it.
I'm closing it. You can take over this if free. @lzmhhh123 @XuHuaiyu

**Pull request closed without merging**